sketch/src/sketch_rnn/model.ts

```diff
@@ -0,0 +1,435 @@
+      }
+      this.weights.push(rawWeights);
+    }
+    this.outputKernel = tf.tensor2d(this.weights[0],
+      [this.weightDims[0][0], this.weightDims[0][1]]);
+    this.outputBias = tf.tensor1d(this.weights[1]);
+    const lstmKernelXH = tf.tensor2d(this.weights[2],
+      [this.weightDims[2][0], this.weightDims[2][1]]);
+    const lstmKernelHH = tf.tensor2d(this.weights[3],
+      [this.weightDims[3][0], this.weightDims[3][1]]);
+    const axis = 0;
+    this.lstmKernel = tf.concat2d([lstmKernelXH, lstmKernelHH], axis);
+    this.lstmBias = tf.tensor1d(this.weights[4]);
+
+    this.rawVars = [
+      this.outputKernel,
+      this.outputBias,
+      this.lstmKernel,
+      this.lstmBias
+    ];
+
+  }
+
+  /**
+   * Loads variables from the JSON model
+   */
+  async initialize() {
+    this.dispose();
+
+    const vars = await fetch(this.checkpointURL)
+      .then((response) => response.json());
+
+    this.instantiateFromJSON(vars[0], vars[1], vars[2]);
+
+    this.initialized = true;
+    console.log('Initialized SketchRNN.');
+  }
+
+  dispose() {
+    if (this.rawVars) {
+      for (let i = 0; i < this.rawVars.length; i++) {
+        this.rawVars[i].dispose();
+      }
+      this.rawVars = undefined;
+    }
+    if (this.forgetBias) {
+      this.forgetBias.dispose();
+      this.forgetBias = undefined;
+    }
+    this.initialized = false;
+  }
+
+  /**
+   * Sets the internal EXTRA factor of this model (pixel to model space)
+   *
+   * @param scale (the extra scale factor for pixel to model space)
+   *
+   * @returns nothing
+   */
+  setPixelFactor(scale: number) {
+    // for best effect, set to 1.0 for d3 or paper.js, 2.0 for p5.js
+    this.pixelFactor = scale;
+    this.scaleFactor = this.info.scale_factor / this.pixelFactor;
+  }
+
+  /**
+   * Updates the RNN, returns the next state.
+   *
+   * @param stroke [dx, dy, penDown, penUp, penEnd].
+   * @param state previous [c, h] state of the LSTM.
+   *
+   * @returns the next [c, h] state of the LSTM`.
+   */
+  update(stroke: number[], state: Float32Array[]) {
+    const out = tf.tidy(() => {
+      const numUnits = this.numUnits;
+      const s = this.scaleFactor;
+      const normStroke =
+        [stroke[0]/s, stroke[1]/s, stroke[2], stroke[3], stroke[4]];
+      const x = tf.tensor2d(normStroke, [1, 5]);
+      const c = tf.tensor2d(state[0], [1, numUnits]);
+      const h = tf.tensor2d(state[1], [1, numUnits]);
+      const newState = tf.basicLSTMCell(
+        this.forgetBias,
+        this.lstmKernel,
+        this.lstmBias,
+        x,
+        c,
+        h);
+      return newState;
+    });
+    const newC = out[0].dataSync();
+    const newH = out[1].dataSync();
+    for (let i = 0; i < out.length; i++) {
+      out[i].dispose();
+    }
+    return [newC, newH];
+  }
+
+  /**
+   * Updates the RNN on a series of Strokes, returns the next state.
+   *
+   * @param strokes list of [dx, dy, penDown, penUp, penEnd].
```

Comment:
Make a `Stroke` object too.

Reply:
Thanks for the suggestion. As discussed in the previous comments, I really want to keep the strokes as a simple array of array, and not as an array of classes/interfaces, or class containing an array of array. This way it is consistent with the sketch-rnn paper (https://arxiv.org/abs/1704.03477)  in the definition of a sketch in Sect 3.1 and Figure A2. I feel it is also simpler for using the data directly when creating code sketches.